// @vue-storefront/shopify-apollo/lib/api/getArticle.js
import { gql } from '@apollo/client/core';
export async function getArticle(context, params, customQuery) {
    const articleQuery = gql `
  query getArticle(
    $id: ID!
  ) {
    node(id: $id) {
      ... on Article {
        id
        handle
        onlineStoreUrl
        publishedAt
        tags
        title
        content
        authorV2 {
          bio
          email
          firstName
          lastName
          name
        }
        image {
          transformedSrc
          altText
          url
        }
        seo {
          description
          title
        }
        blog {
          handle
          title
        }
      }
    }
  }
`;
    const { article } = context.extendQuery(customQuery, {
        article: {
            query: articleQuery,
            variables: params
        }
    });
    const response = await context.client.apolloClient.query({
        query: article.query,
        variables: article.variables
    });
    const node = response?.data?.node;
    return {
        ...response,
        data: {
            ...response?.data,
            article: {
                ...node,
                fullAuthorName: node?.authorV2?.name ?? ''
            },
        }
    };
}
